modules/diskq/qdisk.c

```diff
@@ -230,7 +237,11 @@ qdisk_is_space_avail(QDisk *self, gint at_least)
       if (_is_write_head_less_than_max_size(self))
         return TRUE;
 
-      return _is_able_to_reset_write_head_to_beginning_of_qdisk(self);
+      /* exact size-check is needed as we have unread/unacked data after the write head
+       * that is being reset
+       */
+      return _is_able_to_reset_write_head_to_beginning_of_qdisk(self)
+             && _is_free_space_at_the_beginning_of_qdisk(self, at_least);
```

Comment:
I think `return _is_free_space_at_the_beginning_of_qdisk(self, at_least);` is enough.

My (overcomplicated) proof:

Let A be `_is_able_to_reset_write_head_to_beginning_of_qdisk()` and B be `_is_free_space_at_the_beginning_of_qdisk()`.

Possible cases are:
  * A: true, B: true -> true
  * A: true, B: false -> false
  * A: false, B: true -> false
  * A: false B: false -> false

When checking the implementation of A and B, if we presume that msg_size is 0 or larger then there cannot be a case where A is false and B is true.

If we remove `A: false, B: true -> false` from the cases above, we get this:
  * A: true, B: true -> true
  * A: true, B: false -> false
  * A: false B: false -> false

From this we can see, that the result is true if and only if B is true. So it is enough to only check B.

Reply:
You are absolutely right. The only reason I would rather leave this condition in is that it might help the learning/understanding process of how qdisk works.

(If I'm lucky, I'll even be able to remove all of these pre-checks from the code and replace all position-correction logic with a simple modulo operation in the next PR.)

Reply:
Cool, thanks for thinking it through:)